rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -81,55 +82,58 @@ struct @(spec.base_type.type)_
-@[  end if]@
-  {
-@# generate constructor body
-@[  if alloc_name and not used_alloc_name]@
-    (void)@(alloc_name);
-@[  end if]@
 @{
-for field in spec.fields:
-    # default values of dynamic arrays with allocators if available
-    if alloc_name and field.type.is_array and \
-            (field.type.array_size is None or field.type.is_upper_bound) and \
-            field.type.is_primitive_type() and field.default_value is not None:
-        cpp_value = value_to_cpp(field.type, field.default_value)
-        print('    %s = %s;' % (field.name, cpp_value))
+def init_list(alloc_name):
+    retstring = ":"
+    used_alloc = False
+    for field in spec.fields:
+        # dynamic arrays require allocator if available
+        if field.type.is_array and \
+                (field.type.array_size is None or field.type.is_upper_bound):
+            retstring += ' %s(%s),' % (field.name, alloc_name)
+            used_alloc = True
+
+    # Always remove the last element; it will either be a trailing comma we do
+    # not want, or a colon we do not want.
+    return (retstring[:-1], used_alloc)
+
+# This function generates constructors.  When the `allocator` argument
+# is false, it generates a "bare" constructor with no allocator argument.
+# When it is true, it generates one with an allocator argument.  In both cases
+# it generates a constructor with an init argument set to the default of ::ALL,
+# which initializes all fields in the message.
+def gen_constructor(allocator):
+    alloc_arg = ''
+    initstr = ''
+    used_alloc = False
+    if allocator:
+        alloc_arg = 'const ContainerAllocator & _alloc, '
+        initstr, used_alloc = init_list('_alloc')
+    # When generating a bare constructor, the end result will look something like:
+    #    explicit Foo_(MessageInitialization init = MessageInitialization::ALL)
+    # When generating an allocator constructor, the end result will look something like:
+    #    explicit Foo_(const ContainerAllocator & _alloc, MessageInitialization init = MessageInitialization::ALL)
+    #    : field1(_alloc), field2(_alloc)
+    print('  explicit %s_(%srosidl_generator_cpp::MessageInitialization init = rosidl_generator_cpp::MessageInitialization::ALL)' % (spec.base_type.type, alloc_arg))
+    if initstr:
+        print('  %s' % (initstr))
+    print('  {')
+    if allocator and not used_alloc:
+        print('    (void)_alloc;')
+    print('    if (init == rosidl_generator_cpp::MessageInitialization::ALL) {')
+    print('      this->initialize_all();')
+    print('    } else if (init == rosidl_generator_cpp::MessageInitialization::SKIP) {')
+    print('      // no work to do')
+    print('    } else if (init == rosidl_generator_cpp::MessageInitialization::ZERO) {')
+    print('      this->zero_initialize();')
+    print('    } else if (init == rosidl_generator_cpp::MessageInitialization::DEFAULTS_ONLY) {')
+    print('      this->initialize_defaults_only();')
+    print('    }')
+    print('  }')
```

Comment:
Can this be refactored to actually use the fact that this is a template? The extra indirection of using `print` here make the code even more difficult to read.

Reply:
I went back and forth on this; it wasn't clear to me that using it as a template actually made it easier to read in all cases.  After looking at the implementation again, I agree with you that in the case of `gen_constructor`, it makes more sense to use it as a template and not as a function with a bunch of `print` statements.  I'll change that.  For `gen_init_func` (below), I don't believe that this is the case; it is mostly control structures, with just a handful of prints embedded, and the templated version is going to be very difficult to read.  For now, I'm just going to change `gen_constructor`, and let's go from there.

Reply:
> For `gen_init_func` (below), I don't believe that this is the case; it is mostly control structures, with just a handful of prints embedded, and the templated version is going to be very difficult to read.

Then I would recommend to keep the control structure in a function but move the prints out of the function and do them in em. I am pretty sure the result is more readable.